platform/openide.dialogs/src/org/openide/NotifyDescriptor.java

```diff
@@ -129,24 +130,50 @@ public class NotifyDescriptor extends Object {
      */
     public static final String PROP_INFO_NOTIFICATION = "infoNotification"; // NOI18N
 
+    /**
+     * Used to be a <code>new Integer(JOptionPane.FOO_OPTION)</code> instance.
+     *
+     * For compatibility reasons the public Object constants must have distinct
+     * identity, but equals must behave like Integer's equals and compare only
+     * the wrapped int like a Record.
+     */
+    private record ReturnValue(int option) implements Serializable, Comparable<ReturnValue> {
+
+        @Override
+        public String toString() {
+            return String.valueOf(option);
+        }
+
+        @Override
+        public int hashCode() {
+            return Integer.hashCode(option);
+        }
+
+        @Override
+        public int compareTo(ReturnValue other) {
+            return Integer.compare(option, other.option);
+        }
+
+    }
+
     //
     // Return values
     //
 
     /** Return value if YES is chosen. */
-    public static final Object YES_OPTION = new Integer(JOptionPane.YES_OPTION);
+    public static final Object YES_OPTION = new ReturnValue(JOptionPane.YES_OPTION);
```

Comment:
Before the patch, NotifyDescriptor.YES_OPTION.equals(JOptionPane.YES_OPTION) returns true.
After the patch, it will return false.

Is that a problem?

Reply:
it could be a problem but is likely unavoidable. The API promises to return Object, by comparing it like that, you make the assumption that the object is an Integer, since the right hand side is a unboxed int:

The way to mitigate it would be to let equals of `ReturnValue` accept Integer too. But this breaks the equals symmetry rule, since this can't be implemented for the other way around. This would cause other problems which are likely worse.